### ISSUE TITLE
[Bug] Automation settings not working on desktop or mobile

### ISSUE THREAD

**Issue by AdnanHodzic** [darkreader/darkreader#9903]
# Bug Report

## Website Address

Same problem on every website.

## Bug Description

Automation settings (set active hours/activate after sunset at location) not working on desktop/laptop (Ubuntu Linux & MacOS) or mobile (Android 12). 

## Steps To Reproduce

After setting automation settings to turn Dark Reader on/off during certain period, i.e: from 07:30PM to 07:30AM this settings will not be respected and Dark Reader will never turn on/off automatically (see screenshot for reference).

Same is if I add latitude/longitude coordinates for my location, for "Activate sunset at location").

Since I use 24 hour, I even try setting it to AM/PM, and again my setting aren't picked up automatically by Dark Reader. 

## Expected Behavior

Dark reader will automatically be turned on/off during above set criteria.

## Actual Behavior

No changes in Dark Reader, however if I check/uncheck box in front of "set active hours/activate after sunset at location" Dark Reader will work as expected. Which defies the whole purpose of "automation" setting.

## Screenshots

![alt text](http://foolcontrol.org/wp-content/uploads/2022/09/dark-reader-automation-settings.png)

## System Information

This problem has been ongoing since Firefox 104 on all devices.

**ThinkPad X1**

- OS: Ubuntu 22.04.1
- Firefox: 105.0 (64-bit) - Mozilla Firefox Snap for Ubuntu canonical-002 - 1.0 (stable channel)
- Dark Reader: 4.9.57

**MacBook M1 Pro**

- OS: MacOS Monterey (12.6)
- Firefox: 105.0 (64-bit)
- Dark Reader: 4.9.57

**Samsung Galaxy S20+**

- OS: Android 12
- Firefox: 105.1.0
- Dark Reader: 4.9.57

**Comment by bershanskiy** [comment]
@AdnanHodzic Thanks for the bug report. I tried to reproduce this on Firefox and noticed that:
 - on Windows 11, after Firefox updated from 104 to 105.0.1, Dark Reader automation did not work. I tried to debug and toggled automation OFF/ON and was not able to reproduce the issue anymore.
 - was not able to reproduce on Android 12 Samsung Galaxy S20 FE with One UI 4.1 (it should share the same bugs as your Samsung Galaxy S20+ since it's very similar to yours, just cheaper)

Based on the glitch on Windows 11, as temporary mitigation, I recommend toggling automation ON/OFF once and seeing if automation starts working again.

However, [this comment](https://github.com/darkreader/darkreader/issues/9533#issuecomment-1254590108) suggests this bug also appears on Chrome. Earlier, I observed that Chrome on Mac did not fire `alarm`s (which is the way Dark Reader knows when to turn on/off). These bugs are all very hard to reproduce so can't debug them right now. Please share any additional info (even seemingly minor) since it could help debug this.

**Comment by AdnanHodzic** [comment]
@bershanskiy 

> on Windows 11, after Firefox updated from 104 to 105.0.1, Dark Reader automation did not work. I tried to debug and toggled automation OFF/ON and was not able to reproduce the issue anymore.

As I also mentioned, this will work for me ...

> however if I check/uncheck box in front of "set active hours/activate after sunset at location" Dark Reader will work as expected. Which defies the whole purpose of "automation" setting.

As from what I understand and from what I would expect is that automation settings to automatically turn Dark Reader on/off during selected hours. If not then this setting make any difference between toggling Dark Reader "on/off".

> was not able to reproduce on Android 12 Samsung Galaxy S20 FE with One UI 4.1 (it should share the same bugs as your Samsung Galaxy S20+ since it's very similar to yours, just cheaper).

Yes, you pretty much described it perfectly and they are very similar devices. I just activated automation back on and Dark Reader turned off  automatically at selected hours. I'll keep an eye on if it turns it back on automatically later on tonight. 

> Based on the glitch on Windows 11, as temporary mitigation, I recommend toggling automation ON/OFF once and seeing if automation starts working again.

As I said above, this is the **only** way I get it to work. On desktop it's not as much of a problem not having automation work, as I also have shortcuts to activate it. But this is lot more troublesome on mobile when in middle of night I have to go to "Options > Addons > Dark reader "toggle on/off"  just to get Dark Reader to turn on/off.

> However, https://github.com/darkreader/darkreader/issues/9533#issuecomment-1254590108 suggests this bug also appears on Chrome. Earlier, I observed that Chrome on Mac did not fire alarms (which is the way Dark Reader knows when to turn on/off). These bugs are all very hard to reproduce so can't debug them right now. Please share any additional info (even seemingly minor) since it could help debug this.

Since you're mentioning alarms which are fired, could it be that one of extensions I'm using are blocking the call? 

On mobile I have following Addons enabled:

- uBlock Origin
- Dark Reader
- Google Search Fixer

Since our phones are so similar, could "UBlock Origin" be blocking something?

On laptops (I have same settings/extensions synced), list is much larger:

- uBlock Origin
- Dark Reader
- Bypass Paywalls Clean
- User-Agent Switcher (only enabled for couple of domain names)
- UltraWideo
- Simple Tab Groups
- Simple Translate
- Move Tabs
- Firefox Translations
- Duplicate Tab Shortcut
- Bitwarden
- Auto HD/4k/8k

From this list suspects could "uBlock Origin" and/or "Bypass Paywalls Clean".

Please let me know if I can send you any additional info to help you get this debugged/replicated on your side.

**Comment by bershanskiy** [comment]
> I'll keep an eye on if it turns it back on automatically later on tonight.

Thanks, that would be very helpful.

> Since you're mentioning alarms which are fired, could it be that one of extensions I'm using are blocking the call?

No, alarms happen in an entirely isolated background context and no other extension should be able to interfere with them. However, I did notice some irregularities with the way Chrome (not Firefox) was delivering alarms.

> Please let me know if I can send you any additional info to help you get this debugged/replicated on your side.

Let's see whether Dark Reader on your Firefox changes the color scheme correctly. Perhaps, it is a bug that occurs only after the update and then is fixed by manually toggling automation ON/OFF once.

**Comment by AdnanHodzic** [comment]
Summary of my experience for past week is that Dark Reader automation setting just doesn't work and if it does it's very inconsistent. I can with almost absolute certainty confirm that after hours are changed/updated, for settings to be applied checkbox should be re-checked each time. 

Best results I've got are on my Android device, where it worked almost perfectly but then there was a day when Dark Reader would just not turn on/off automatically. On Linux it worked the worst, where just now I had to manually turn it on again. On Mac, I would say settings were working 50% of the time. 

Hence it's very hard to isolate and pinpoint what exactly is happening and what's causing it for sometimes to work,

**Comment by PeterStindberg** [comment]
I have the same issue for quite a while now. Sometimes it works, most times it works not. Checking/unchecking it gets the desired result every time. MacBook pro on Mojave with Chrome.

**Comment by bershanskiy** [comment]
I did some more testing and here is a summary of relevant information:
 1. Dark Reader detects value of color scheme and change in it via [`prefers-color-scheme`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme) media query
 2. Firefox had been previously reported as not delivering media query change events after color scheme changes in _some_ tabs while delivering them in _other_ tabs (I observed it on Windows and another user on MacOS)
 3. I'm observing Chrome reporting incorrect (light) color scheme and not delivering media query change events right now on Linux (Ubuntu 22.04, default DE settings). This happens in _all_ tabs.
 4. Firefox incorrectly reports color scheme preferences in extensions' background contexts after background startup
 5. To work around 2, Dark Reader sets up media query change listeners in multiple contexts (so that Dark Reader has a greater _probability_ of being notified of the change). These contexts include the extension background and every a content script of active tabs.
 6. I'm not aware of any way to work around 3.
 7. To work around 4, Dark Reader ignores initial color scheme in Firefox background context.
 8. I had tried to toggle dark mode on and off repeatedly to try to replicate the bug and on Firefox I observed Dark Reader icon "blinking". Repro steps:
    - open multiple tabs, all on different domains; set Dark Reader to system automation
    - Repeatedly do the following:
       - Set system to dark (extension is OFF)
       - Toggle system to dark
       - DR icon changes to dark for a moment
       - DR icon changes to light for a moment
       - DR icon changes to dark and remains dark

Point 8 might indicate that Firefox delivers incorrect color preferences in some color change events. Sometimes, the last received event has the correct value (and then icon just blinks and user does not notice this). Sometimes, the last received event might have incorrect value and then Dark Reader automation breaks in a noticeable way. I'm going to look into this bug a bit more, and I would appreciate any extra information you might have.

**Comment by bershanskiy** [comment]
I did a bit more testing on Firefox and found that occasionally Firefox reports old value in some contexts. For example, in the following screenshot (a portion of longer log) I would have expected to see long sequences of `true` followed by long sequences of `false`. Instead, I see a lone `false`. It could be a stale value or change event delivered very late (after color changed again). I don't know.
![image](https://user-images.githubusercontent.com/45960703/193582958-0510db9a-8b90-450a-a4b8-9801094eee94.png)

However, we can:
1. Listen for change event in all contexts we currently listen in, but ignore the actual reported value
2. When we are notified of change, request the actual value from the main frame of the active and visible tab, if there is one. This tab is the most likely to have the correct value (since browsers delay notifying optimiz
3. Since 2 would produce a lot of repeated messaging, we would need to debounce the check and perform it no more than, say, once every 200 ms.

@Gusted What do you think?

**Comment by bershanskiy** [comment]
Of course, the above two comments apply only to "System" option.

Regarding "time" option: I observed "time" not working before but I could never reproduce it reliably. Specifically, I saw once that Chrome stooped delivering these alarms. I'm not sure how we would fix that, aside from manually checking scheduled alarms during processing of some unrelated event.

**Comment by Gusted** [comment]
> I did a bit more testing on Firefox and found that occasionally Firefox reports old value in some contexts. For example, in the following screenshot (a portion of longer log) I would have expected to see long sequences of `true` followed by long sequences of `false`. Instead, I see a lone `false`. It could be a stale value or change event delivered very late (after color changed again). I don't know.

Could we make this into a small extension? Such we can report it to Firefox.

> 
    Listen for change event in all contexts we currently listen in, but ignore the actual reported value
    When we are notified of change, request the actual value from the main frame of the active and visible tab, if there is one. This tab is the most likely to have the correct value (since browsers delay notifying optimiz
    Since 2 would produce a lot of repeated messaging, we would need to debounce the check and perform it no more than, say, once every 200 ms.

I'm fine with having the listener in every frame possible(and hope that doesn't cause any issue). And on the background if we have such message(that we receive from tabs), we will send a message to the active tab frame and use their value and assume that's the "correct" one.

**Comment by Gusted** [comment]
> Regarding "time" option: I observed "time" not working before but I could never reproduce it reliably. Specifically, I saw once that Chrome stooped delivering these alarms. I'm not sure how we would fix that, aside from manually checking scheduled alarms during processing of some unrelated event.

Once again, let them fix their problem (if we can make it reproducible).

**Comment by ildar** [comment]
> Once again, let them fix their problem (if we can make it reproducible).

Is there a ref to the issue at Mozilla?

**Comment by ildar** [comment]
I personally think it may be related to the suspend/resume cycle. Does the extension track that?

**Comment by blendmaster** [comment]
I have this same problem, on firefox stable 105, with "switches according to specified time" automation. I did find that if I exit firefox completely (swipe away in the app switcher), then reopen firefox, the correct automation applies.

Possibly unrelated issue: when I do change Dark Reader's own operating mode from "auto" to either "on" or "off" then switch back to automation, the automation mode changes to "switches according to system dark mode", not retaining my previous setting to switch by time of day. I can change the automation settings back, but it is annoying the setting isn't preserved.

**Comment by bershanskiy** [comment]
> Possibly unrelated issue: when I do change Dark Reader's own operating mode from "auto" to either "on" or "off" then switch back to automation, the automation mode changes to "switches according to system dark mode", not retaining my previous setting to switch by time of day. I can change the automation settings back, but it is annoying the setting isn't preserved.

This is separate issue. I think this behavior was intentional, but I'm open to changing it (preserving the preferred automation mode), I'll discuss it with others and get back to you.

**Comment by PeterStindberg** [comment]
I ight have a tidbit that might help: I run Dark Reader on Chrome (see specs above in the thread), but I run the browser under TWO profiles. One profile manages the switch, the other does not. (and yes, the plugin is on both of course)

**Comment by PeterStindberg** [comment]
fwiw - I switched to Firefox in the meantime, and the problems persist. On most days, the settings get ignored. On some days, it switches one way, but not the other. I have in the menatime resorted to setting the settings manually.

**Comment by tony1661** [comment]
I'd also like to report that I am experiencing this issue on Linux Mint 21 as well as the latest PopOS. This issue for me is both on Chrome and Firefox.

**Comment by janxkoci** [comment]
> Based on the glitch on Windows 11, as temporary mitigation, I recommend toggling automation ON/OFF once and seeing if automation starts working again.

The problem with this workaround is that I have to do this literally every day :confused: . I use DarkReader in Firefox on my Samsung A20e phone as well as on a Linux laptop (elementary OS 6).

I tried automation based on time, or system theme, but nothing works reliably - I need to fiddle with the settings pretty much every day to get it back working. If at least I didn't have to go through 3 screens to get there, and then the same 3 screens back (on phone - this used to be one tap before).

**Comment by amit-batra** [comment]
Ever since the automation feature got broken in the Dark Reader extension, I noticed that:

1. All 3 types of automation (time based, system theme based, sunset/sunrise based) are broken.
2. It's broken on both Windows and on macOS.
3. It's broken on Firefox, Brave and Chrome browsers.

There might be other platforms and browsers affected besides the ones that I have listed, but my experience is limited to the ones above.

**Comment by janxkoci** [comment]
@amit-batra You can add Firefox on Android to the list..

**Comment by vagra** [comment]
![图片](https://user-images.githubusercontent.com/1576687/226773402-d303c2c1-4711-400f-8f51-ad88f68df9ec.png)

Windows 10 
Firefox 111
Edge 111

darkreader 4.9.62

no (Toggle dark/light) options.

**Comment by xrad** [comment]
> I personally think it may be related to the suspend/resume cycle. Does the extension track that?

I have the same suspicion. Could it be that suspend/resume simply cancels pendig alarm listeners?

**Comment by janxkoci** [comment]
As I experienced this _minutes_ after cold boot, I don't think these are related.

**Comment by Arzte** [comment]
I have noticed that sometimes it just takes a few minutes for dark reader to realize the time has changed, it seems to *usually* turn on properly if I set it to turn on 15 minutes+ before I actually want it on.

I'm currently on Firefox 112.0.1, Dark Reader 4.9.63, and macOS 13.2.1.

**Comment by janxkoci** [comment]
> if I set it to turn on 15 minutes+ before I actually want it on

The time of day when I need it "on" changes progressively throughout the year - I'd generally need to tweak it every few weeks (I know, because it's what I did before automation was a thing, or with tools where it still doesn't work). That's a PITA, if you ask me (or anyone who'd done this long-term).

**Comment by Arin-426** [comment]
Is no one facing this issue anymore? Automation is still not working for me, Brave browser. Same on Firefox and Chrome.

Can someone please suggest a workaround or solution to this?

**Comment by ildar** [comment]
It started to work normally at some point. I think it's when I switched to
Hyprland + xdg-portal. FF.

**Comment by Arin-426** [comment]
So am I the only now who's facing the issue? :/

I'm on Windows 10. Does anybody know the cause?

**Comment by Arin-426** [comment]
> @Arin-426 I don't have this issue anymore for months, maybe a year, on both desktop (Linux) & mobile (android) Firefox.

@janxkoci Hey, I figured it out. Apparently, it's the mv3 version that Google is stupidly forcing over us. I downloaded the mv2 extension from the repo here, and everything works perfectly (except the context menu dark mode, maybe I am missing something).

**Comment by janxkoci** [comment]
@huachuman does Vivaldi use the new extension standard (mv3) as mentioned in the comment above yours? Perhaps you can try the other version of the extension too..

**Comment by huachuman** [comment]
Vivaldi is sticking with V2 until 2025

**Comment by janxkoci** [comment]
Okay, and have you tried the V2-compatible extension from this repo, as described above?
